(closes 2204) Added eqv support in Fortran backend and enabled fparser frontend to …

LonelyCat124 opened the pull request:
…read eqv from Fortran input

This is the option 1 solution i proposed I think. I think the change to `get_operator` is a little ugly but it does work - any alternative proposals would be good.

As discussed, this won't handle logical expression comparisons, only Literals and References.

It should be tested as well.

LonelyCat124 commented:
@arporter @rupertford I might need a bit of help diagnosing why the example is failing here, I have this example failing and I can't see why:
```
(LFRIC_env) [achalk@glados tangent_linear]$ psyad -api dynamo0.3 helmholtz_operator_kernel_mod.F90 -oad ../adjoint_partial/adj_helmholtz_operator_kernel_mod.F90 -a helm_c f_op helm_dd helm_uu helm_d helm_u helm_s helm_n helm_w helm_e d_op ec_op m3_exner_star p3theta compound_div
The array index expressions for access 'smap_w2(d2,1 + dir)' include a function call or expression. Querying the return type of such things is yet to be implemented.
```

I hit this error in `array_mixin`:
```
            elif isinstance(idx_expr, (Call, Operation, CodeBlock)):
                # We can't yet straightforwardly query the type of a function
                # call or Operation - TODO #1799.
                raise NotImplementedError(
                    f"The array index expressions for access "
                    f"'{self.debug_string()}' include a function call or "
                    f"expression. Querying the return type of "
                    f"such things is yet to be implemented.")
```

where `idx_expr` is the `1 + dir` - I'm not clear how many changes have caused this?

LonelyCat124 commented:
I ran the tests on glados and all 5216 passed so there's something missing I guess that is hit by that example somehow :(

arporter commented:
> where `idx_expr` is the `1 + dir` - I'm not clear how my changes have caused this?

It's because you are now querying the datatype of things that appear on either side of a `==`:

      do d2 = 1, nfaces_h
        if ( smap_w2(d2,1+dir) == smap_w2(dir,1) ) adjacent_face(dir) = d2
      end do
    end do

Since we also check for array shape (rather than just intrinsic type), we examine the indices of the access and at that point we encounter an expression (albeit a very simple one) and abort.
I suggest you examine the type of the Symbol (that the Reference is to) directly to avoid this problem for now.

LonelyCat124 commented:
Thanks Andy, I've switched to that and I think the examples and tests should all build now. If they do I'll mark this as ready for review.

LonelyCat124 commented:
Ready for review now.

sergisiso commented:
I am a bit worried with this solution, isn't the new
`('.eqv.', BinaryOperation.Operator.EQ)` in fparser
going to make parsing Fortran like:
`if (a.and.b .eqv. c.and.d) then`
go from valid Fortran code to invalid code because we would generate the ==?

It's worst because we don't always have the types, e.g:
```
use mod, only: a, b
if (a .eqv. b) then
```
we have no way to reproducing this code without chasing the dependencies.

Is this right?

If that's the case I think I would prefer your second option of adding EQV in PSyIR BinaryOps.
It would still allow to use EQ with logical but not having typechecking is an issue consistent with the rest of PSyIR
But more importantly Fortran->PSyIR->Fortran won't introduce bugs.

LonelyCat124 commented:
It will I believe yes, which is not good, but this is an improvement on the current situation for simpler cases (the code fails to be read at all). It also fixed `SELECT CASE` statements producing incorrect code if the cases are .true. and .false. - so we trade one case of incorrect output for another.

 I think the future solution we originally discussed was that expressions will be able to be handled by a future PR that @arporter is planning to do for `ASSOCIATE`, so we will leave a TODO in the code and it will be updated when we can check the type of a general expression (if I understood correctly).

This wouldn't resolve the issue for unknown type problem though I think there is always an issue when parsing `SELECT CASE`, for example:
```
use external_mod, only: a, b, c

select case(a)
case(b)
   call x
case(c)
   call y
end select
```
PSyclone will not know whether to use `eqv` or `EQ` at construction of the if statements in `fparser2.py` here I think. There may not be a solution that works for all cases.

I do think that perhaps `EQV` as a separate operator might be better to preserve original syntax though - I am not sure how to transform `_process_case_value` to fix the errors I had in #2204, can I do something like I have added to the fortran backend, but apply it to `bop` after doing the `process_nodes` call and change the operator of `bop` if it is determined it should be `EQV`?

@arporter @rupertford Does this seem reasonable to you?

arporter commented:
A more conservative (small 'c') solution is to have the frontend reject situations where we can't be sure whether .EQV. is needed and just put them in a CodeBlock.

LonelyCat124 commented:
@sergisiso @arporter I think I swapped the code around to now make .eqv. the frontend's problem as Sergi recommended.
This should deal with some of the cases Sergi was concerned would fail, and I believe the Select Case will work correctly for all Select Case blocks now. The ones that we don't understand will result in a `CodeBlock`.

I think I'm missing some tests for `.eqv.` passing through the system without being touched correctly, but I was not sure which file should contain them as `psyir/frontend` test files didn't seem to have a set of tests for testing binary operation construction - any ideas?

sergisiso commented:
Thanks @LonelyCat124, I think this is the right approach. The BinaryOperator is tested in `src/psyclone/tests/psyir/frontend/fparser2_test.py::test_handling_binaryopbase` and EQV must be added to that list (but note that we bundle together arithmetic and logic operators), maybe the logic operators have to be split into a second equivalent test?

LonelyCat124 commented:
I added to the existing test for now, and think this is ready for review.

sergisiso reviewed (changes requested):
Thanks @LonelyCat124 the PR looks good. I am just asking if .nevq. and .xor. could also be added doing the same so we can process full logical expressions in the future and some other minor code/test suggestions.

LonelyCat124 commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -48,7 +48,8 @@
     TYPE_MAP_FROM_FORTRAN
 from psyclone.psyir.nodes import BinaryOperation, Call, CodeBlock, DataNode, \
-    IntrinsicCall, Literal, Operation, Range, Routine, Schedule, UnaryOperation
+    IntrinsicCall, Literal, Operation, Range, Routine, Schedule, \
+    UnaryOperation, Reference
```
Done

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -3234,9 +3237,48 @@ def _process_case_value(self, selector, node, parent):
             # The case value is some scalar initialisation expression
```
Removed

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -3234,9 +3237,48 @@ def _process_case_value(self, selector, node, parent):
-            parent.addchild(bop)
             self.process_nodes(parent=bop, nodes=[selector])
             self.process_nodes(parent=bop, nodes=[node])
+            # TODO #2204 This will not support if we have two expressions
+            # or a unknown type, as we cannot determine if we have == or EQV
+            # operator for thoses cases. This should result in a CodeBlock
+            # Keep track of if we know if the operator should be EQ or EQV
```
Done

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -3234,9 +3237,48 @@ def _process_case_value(self, selector, node, parent):
+            if operator_known:
+                parent.addchild(bop)
+            else:
+                raise NotImplementedError("PSyclone can't determine if this "
+                                          "case should be == or .EQV. so "
```
Done

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -1013,6 +1012,7 @@ class Fparser2Reader():
         ('**', BinaryOperation.Operator.POW),
         ('==', BinaryOperation.Operator.EQ),
         ('.eq.', BinaryOperation.Operator.EQ),
+        ('.eqv.', BinaryOperation.Operator.EQV),
```
I added NEQV and XOR did some tests for the precedence of them. Notably XOR is a "legacy-extension", so fparser can't parse it at the moment. I've left it in in case we decide to support it in future but I can remove it if you'd prefer.

LonelyCat124 commented on src/psyclone/tests/psyir/frontend/fparser2_select_case_test.py:
```diff
@@ -0,0 +1,175 @@
+from __future__ import absolute_import
+
+from psyclone.psyir.nodes import CodeBlock
+
```
Done. I couldn't see any other select/case related tests.

LonelyCat124 commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -2760,6 +2758,24 @@ def test_handling_binaryopbase():
         assert fake_parent[0].rhs._operator == expected, \
             "Fails when parsing '" + opstring + "'"
 
+    # Test parsing all supported logical binary operators.
```
I added to the precedence test for NEQV, EQV and XOR but only NEQV and EQV for the fw test as XOR can't be parsed yet.

LonelyCat124 commented on src/psyclone/tests/psyir/frontend/fparser2_select_case_test.py:
```diff
@@ -0,0 +1,175 @@
+
+
+def test_unknown_types_case(fortran_reader):
+    '''Test that a seclect case statement comapring two unknown types
```
Done

LonelyCat124 commented:
@sergisiso Fixed up the requested changes, note that .xor. is an extension, so while PSyclone can handle it, fparser currently can't. I can therefore remove the xor changes if you'd prefer.

sergisiso commented:
> note that .xor. is an extension

Ah thanks, if its not standard Fortran lets remove it.

LonelyCat124 commented:
@sergisiso done, ready for review now I believe.

sergisiso reviewed (approved):
@LonelyCat124 All comments have been addressed. All testing and formatting checks pass.  No TODOs to this issue remaining. This is ready to merge.
